README.rst
MicroPython micro:bit REPL
==========================

This script lets you run commands directly onto a connected micro:bit device.
Put simply, you get to program the micro:bit in real time, just like we used
to with the old BBC micro from the 1980's. This encourages exploration,
experimentation and a sense of adventure. It's also a good way to play with
the device to learn about its capabilities.

The script detects the port to which the micro:bit is connected and uses
PySerial to make a connection to the micro:bit's Read Evaluate Print Loop
(REPL).

For more information about what a REPL, is check out this Wikipedia article:

https://en.wikipedia.org/wiki/Read%E2%80%93eval%E2%80%93print_loop

Installation
------------

All currently supported Windows versions (8.1+) should automatically install
the micro:bit serial/COM drivers when the device is first plugged in.
All versions of Linux and macOS should work out of the box.

You can install/run the script in three ways:

1. Install via pip and PyPI::

    # pip install microrepl

2. Install via setup.py::

    # python setup.py install

3. Run the script from within the repository::

    # pip install -r requirements.txt
    # python microrepl.py

All of the above should probably be done within a virtualenv. The first two
options will result in two commands being installed: microrepl or its alias
urepl. Simply type these at the command prompt to run the application.

Getting Started
---------------

When you run the script it will try to be as helpful as possible in working out
why it can't connect. In most cases it will try to print some helpful advice.

Once connected the script will stop the program running on the micro:bit in
order to drop you into the REPL itself. You type Python commands next to the
prompt (``>>>``).

If things don't seem to be working, re-flash your micro:bit using the
``firmware.hex`` file in this directory, and try again.

You can use the TAB key to auto-complete words. For example, if you
type ``microbit.sc`` then hit TAB, MicroPython will helpfully complete the
word for you, like this: ``microbit.screen``.

You can move forwards and backwards through your command history by using the
up arrow key (to move back through your command history) and down arrow key
(to move forwards through your command history). The left and right arrow
keys allow you to move through the text of your command without deleting it.

Use the TAB and arrow key commands! They save a lot of typing and look really
cool.

Unplugging the device and pressing the reset button will re-start the program
as the version you originally flashed onto the device.

Useful Commands
---------------

A good place to start is to type ``help()``. From here you should be able
to explore the device and the MicroPython version of Python 3.

All the micro:bit's hardware is available to program via the ``microbit``
module. To start using it type the command ``import microbit``. Do fun things
such as ``microbit.display.scroll('Hello, World!')`` and watch the device react
in real-time.

Type the command ``dir()`` to see a list of all the things currently
available to you - this includes the various classes and functions for the
program you may have already flashed onto the micro:bit.

Find Out More
-------------

To learn more about the amazing MicroPython visit: http://micropython.org/

The BBC micro:bit's (non-Pythonic) home is: http://microbit.co.uk/

To learn about the Python language visit: http://python.org/


CHANGES.rst
Release History
---------------

0.7
+++

Fixed bug #4 when using PySerial version >= 3.0
Fixed issue #16 running on Windows.
Set pyserial minimum version to 3.0.
Removed mbedWinSerial_16466.exe driver as is not necessary in the currently
supported Windows versions.
Set the exit code to 1 when microrepl doesn't find a micro:bit serial port. 

0.6
+++

Fixed bug #13 on OSX

0.5
+++

Updated to latest version of PySerial.

0.4
+++

Windows support added. Requires installation of drivers found in the
mbedWinSerial_16466.exe executable. More information can be found at the ARM
website: https://developer.mbed.org/handbook/Windows-serial-configuration

0.1
+++

Initial release. Works on Linux and OSX on both Python 2.7 and 3.3+. Windows
version coming soon.


tests.py
import pytest
import io
from unittest.mock import patch
import os
import tempfile

from microrepl import connect_miniterm

@pytest.yield_fixture
def fake_stderr():
    fake_stderr = io.StringIO()
    with patch('sys.stderr', fake_stderr):
        yield fake_stderr

@pytest.yield_fixture
def fake_sys_exit():
    with patch('sys.exit', autospec=True) as fake_exit:
        yield fake_exit


def test_connect_miniterm_suggests_solution_to_perms_problem_on_linux(fake_stderr, fake_sys_exit):
    nonaccessible_port = tempfile.NamedTemporaryFile()
    os.chmod(nonaccessible_port.name, 0o000)
    connect_miniterm(nonaccessible_port.name)
    error_message = fake_stderr.getvalue()
    assert "Found micro:bit, but could not connect." in error_message
    assert "[Errno 13] could not open port" in error_message
    assert "Permission denied: {port!r}".format(port=nonaccessible_port.name) in error_message
    assert 'On linux, try adding yourself to the "dialout" group' in error_message
    assert 'sudo usermod -a -G dialout <your-username>' in error_message


run_tests.sh
./.run_tests.py -s tests.py


microrepl.py
#!/usr/bin/env python
"""
A simple shim around PySerial that detects the correct port to which the
micro:bit is connected and attempts to make a serial connection to it in order
to bring up the Python REPL.
"""
from __future__ import print_function
import re
import sys
import serial
import serial.tools.miniterm
from serial.tools.list_ports import comports
from serial.tools.miniterm import Console, Miniterm, key_description


MICROBIT_PID = 516     # 0x0204
MICROBIT_VID = 3368    # 0x0D28
BAUDRATE = 115200
PARITY = 'N'
ARM = 'https://developer.mbed.org/handbook/Windows-serial-configuration'
EXIT_CHAR = chr(0x1D)    # GS/CTRL+]

# Regular expression to match the device id of the micro:bit
RE_VID_PID = re.compile("VID:PID=([0-9A-F]+):([0-9A-F]+)", re.I)


def find_microbit():
    """
    Returns the port for the first micro:bit found connected to the computer
    running this script. If no micro:bit is found, returns None.
    """
    for port, desc, opts in comports():
        match = RE_VID_PID.search(opts)
        if match:
            vid, pid = match.groups()
            vid, pid = int(vid, 16), int(pid, 16)
            if vid == MICROBIT_VID and pid == MICROBIT_PID:
                return port
    if sys.platform.startswith('win'):
        # No COM port found, so give an informative prompt.
        sys.stderr.write('Have you installed the micro:bit driver?\n')
        sys.stderr.write('For more details see: {}\n'.format(ARM))
    return None


def connect_miniterm(port):
    try:
        ser = serial.Serial(port, BAUDRATE, parity=PARITY, rtscts=False, xonxoff=False)
        return Miniterm(
            ser,
            echo=False,
            #convert_outgoing=2,
            #repr_mode=0,
        )
    except serial.SerialException as e:
        if e.errno == 16:
            # Device is busy. Explain what to do.
            sys.stderr.write(
                "Found micro:bit, but the device is busy. "
                "Wait up to 20 seconds, or "
                "press the reset button on the "
                "back of the device next to the yellow light; "
                "then try again.\n"
            )
        elif e.errno == 13:
            print("Found micro:bit, but could not connect.".format(port), file=sys.stderr)
            print(e, file=sys.stderr)
            print('On linux, try adding yourself to the "dialout" group', file=sys.stderr)
            print('sudo usermod -a -G dialout <your-username>', file=sys.stderr)

        else:
            # Try to be as helpful as possible.
            sys.stderr.write("Found micro:bit, but could not connect via" +
                             " port %r: %s\n" % (port, e))
            sys.stderr.write("I'm not sure what to suggest. :-(\n")
        sys.exit(1)


def main():
    """
    The function that actually runs the REPL.
    """
    port = find_microbit()
    print('port', port)
    if not port:
        sys.stderr.write('Could not find micro:bit. Is it plugged in?\n')
        sys.exit(1)
    miniterm = connect_miniterm(port)
    # Emit some helpful information about the program and MicroPython.
    shortcut_message = 'Quit: {} | Stop program: Ctrl+C | Reset: Ctrl+D\n'
    help_message = 'Type \'help()\' (without the quotes) then press ENTER.\n'
    sys.stderr.write(shortcut_message.format(key_description(EXIT_CHAR)))
    sys.stderr.write(help_message)
    # Start everything.
    miniterm.exit_character = EXIT_CHAR
    miniterm.set_rx_encoding('utf-8')
    miniterm.set_tx_encoding('utf-8')
    miniterm.start()
    miniterm.serial.write(b'\x03')  # Connecting stops the running program.
    try:
        miniterm.join(True)
    except KeyboardInterrupt:
        pass
    sys.stderr.write('\nEXIT - see you soon... :-)\n')


if __name__ == '__main__':
    main()
